nodejs/lib/util/webstorage.js

```diff
@@ -0,0 +1,109 @@
+"use strict";
+var WebStorage = (function() {
+	var path = require('path');
+	var os = require('os');
+	var mkdirs = require('node-mkdirs');
+	var JSONStorage = require('node-localstorage').JSONStorage;
+
+	function storedItem(value, ttl) {
+		var item = {
+			value: value
+		};
+		if(ttl) {
+			item.expires = (Date.now() + ttl);
+		}
+		return item;
+	}
+
+	function valid(item) {
+		var expires = item.expires;
+		return !expires || (Date.now() < expires);
+	}
+
+	function WebStorage(client) {
+		this.client = client;
+		this.storagePath = null;
+		this.localStorage = null;
+		this.sessionStorage = new Map();
+	}
+
+	WebStorage.prototype.supportsLocal = true;
+	WebStorage.prototype.supportsSession = true;
+
+	WebStorage.prototype.initPath = function() {
+		var storagePath = this.storagePath;
+		if(!storagePath) {
+			var options = this.client.options, appId;
+			if(options.key) {
+				appId = options.key.split('.')[0];
+			} else {
+				var authOptions = this.client.auth.authOptions,
+					token = authOptions.tokenDetails.token;
```

Comment:
> `authOptions.tokenDetails.token`

This will crash if the client is using authurl/authcallback, as authOptions.tokenDetails won't be available yet.

You could delay Storage init for node until the first time the client successfully authorizes if we're quite sure that that the storage api won't be used by push until after that's happened. Pretty nasty though.

I haven't been following the push conversation, but what's the issue with having a device's device id used across multiple apps (i.e. if we used a fixed path here)? After all, it is ultimately the same device.

Reply:
> You could delay Storage init for node until the first time the client successfully authorizes if we're quite sure that that the storage api won't be used by push until after that's happened. Pretty nasty though.

But `init()` is only called when `set()` is called, so whta's here will work for connection persistence. We could have `init()` force authorisation if it's not already happened, or we can have the push implementation force authorisation ahead of attempting to set any persisted values.

> I haven't been following the push conversation, but what's the issue with having a device's device id used across multiple apps (i.e. if we used a fixed path here)? After all, it is ultimately the same device.

A `deviceId` belongs to an app, and the activation state, which is persisted, depends on the realtime state, which is persisted for the `appId/deviceId` combination.

Aside from that, we do still want to be able to run tests by creating a new app and know that the state of client and server are both clean at that point.

Reply:
> But init() is only called when set() is called,

No it isn't? `this.storage = new WebStorage(this);` is called from the Rest client constructor

Reply:
Yes, but the `WebStorage` constructor doesn't call `init()`.

Reply:
Fair enough. I saw it was crashing the tests and assumed that it was because it was being initialized too early, but looking closer, it's actually it was because it's looking at `auth.authOptions.tokenDetails`; that should be `auth.tokenDetails` (the token is only an authOption if the library was initialized or authorize()d with a token literal, as opposed to having fetched one with authUrl/authCallback).